### PR TITLE
Quick Start for Existing Users V2 - Refactor code to support different quick start types

### PR DESCRIPTION
**Pull request by ashiagr**
Parent #16350

Corresponding FluxC PR: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2384

This PR refactors quick start logic to consider tasks and task types based on quick start type. 
Note 1: No new functionality is added in this PR. It needs a careful code review and regression testing.
Note 2: I've tried to split changes into several commits and provided an overview of the changes in our sync call but let me know if I can help in reviewing the PR.

----

## Details

Changes can be grouped under the following heads:

### QuickStartRepository.quickStartType

Currently, the same set of tasks is shown irrespective of whether the user creates a new site or chooses an existing site. These tasks are currently part of `NewSiteQuickStartType` which is set as qs type in the qs repo in a73828b4.

### Trackers and Illustrations

Updated to compare task strings in trackers and task illustrations as a common task like `check_stats` needs the same tracking, and illustrations irrespective of quick start type. 

Class | Functions | Commits
-- | -- | --
QuickStartItemBuilder | getIllustrations | b51b0074
QuickStartUtils | getQuickStartListTappedTracker, getQuickStartListSkippedTracker, getTaskCompletedTracker | c695d3bb

### QuickStartMySitePrompts, QuickStartNoticeDetails, QuickStartTaskDetails

Updated to get the corresponding enum by comparing task strings.

As a common task like `check_stats` needs the same prompt, notice details, task details, and illustrations irrespective of quick start type, there was no need to split them based on quick start type. 

Class | Commits
-- | --
QuickStartMySitePrompts | 5b3bbc9a
QuickStartNoticeDetails | 001da950..230e50b5
QuickStartTaskDetails | 32b78aa1..2f8a124e

### Quick Start Functions

Class | Functions |   | Commits
-- | -- | -- | --
QuickStartUtilsWrapper | startQuickStart |   | 161d11ae
  | completeTaskAndRemindNextOne |   | 370ee218
QuickStartUtils | isQuickStartInProgress |   | 77d9abaa
  | getNextUncompletedQuickStartTask | Replaced use of all  task types with task types for quick start type | 82abd92f
  | isEveryQuickStartTaskDone | Replaced use of all tasks with  list of quick start type tasks | c61db465
  | getNextUncompletedQuickStartTaskForReminderNotification | Replaced use of all  task types with task types for quick start type | c265b94d
AppPrefs | getLastSkippedQuickStartTask | Get last skipped task from list of quick start type tasks | 62fd05e4
QuickStartRepository | getQuickStartTaskTypes | Replaced hardcoded task types with task types for quick start type | f7c15840
QuickStartReminderReceiver |   | Updated fail safe conditions | 3c83c960

### Quick Start Card

Class | Notes | Commit
-- | -- | --
QuickStartCardViewHolder | Updated get to know app task item type | 8c05d846

----

#### Test.1 Regression test Quick Start 

1. Launch the app.
2. Logout/login and select "Show me around" when the quick start prompt is shown.
3. Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).

#### Test.2 Dynamic Cards

1. Launch the app.
2. Go to My Site -> Me -> App Settings -> Debug settings.
3. Turn on the QuickStartDynamicCardsFeatureConfig flag under the Features in development section.
4. Restart the app.
5. Logout/login and select "Show me around" when the quick start prompt is shown
6. Notice that quick start dynamic cards are shown in My Site tab
7. Make sure that dynamic card interactions are the same as before.

PS: Test 5  - Quick Start Cards disappear in `wpandroid-quick-start-testing-instructions.txt` will not work as `Plans` were removed (internal ref: p1651587009445519-slack-C011BKNU1V5) but the corresponding Quick Start task remains which will not mark the card as complete.

#### Test.3 (optional) To test that qs functions consider tasks and task types based on quick start type. 
(I added this test to share the patch I used to find out qs functions we needed to update based on quick start type but it can be skipped to a future PR when we actually add the tasks)

1. Apply [patch](https://cloudup.com/cgjjIQkUmQb) that sets `ExistingSiteQuickStartType` in qs repo and adds common tasks `check_stats`, `view_site` temporarily in `QuickStartStore`. Enable local fluxc in local-build.gradle.
2. Reinstall app.
3. Logout/login and select "Show me around" when the quick start prompt is shown.
4. Run tests in [wpandroid-quick-start-testing-instructions.txt](https://github.com/wordpress-mobile/WordPress-Android/files/8246096/wpandroid-quick-start-testing-instructions.txt).

---
## Merge Instructions

- Wait for the corresponding FluxC PR to be merged
- Update `FluxC` version in `builds.gradle`
- Remove `Not Ready for Merge` label
- Merge the PR

## Regression Notes
1. Potential unintended areas of impact
Quick Start functionality does not work as expected

2. What I did to test those areas of impact (or what existing automated tests I relied on)
See `To Test` section

3. What automated tests I added (or what prevented me from doing so)
Updated existing tests for the code to compile after refactor (no changes to existing functionality in this PR).

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by jostnes** [comment]
Hey @ashiagr, thanks for the details PR description and testing notes, those are very helpful! Some notes from testing:
1. I noticed one thing after logging in, there seems to be a slight delay when displaying the Quick Start prompt. Notice that the Home screen is displayed briefly before the prompt:

https://user-images.githubusercontent.com/17252150/167761321-fedc5202-19cb-4186-930c-d00ae3fae06b.mp4

Note: Spoke to @ashiagr offline about this, this is from [this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/16435). I also checked the current beta build and the same is reproducible there, so isn't related to the PR. 

2. As mentioned in the description, Plans task on Quick Start can't be completed due to [this issue](https://github.com/wordpress-mobile/WordPress-Android/issues/16477) so Quick Start completed flow is excluded from testing
3. Tested these 4 combinations and overall works as expected (observations on the next point):
- From `Home` tab with `QuickStartDynamicCardsFeatureConfig` turned off
- From `Home` tab with `QuickStartDynamicCardsFeatureConfig` turned on
- From `Menu` tab with `QuickStartDynamicCardsFeatureConfig` turned off
- From `Menu` tab with `QuickStartDynamicCardsFeatureConfig` turned on
4. Some observations from testing: 
- Noticed that user is still not guided to the correct tab (from Menu to Home) in this build. The issue looks to be fixed in [this PR](https://github.com/wordpress-mobile/WordPress-Android/pull/16462). Was wondering if the fixed is included in the PR or if this is a regression?
- Noticed that there is no quick start prompt when the `QuickStartDynamicCardsFeatureConfig` flag is turned on (it will automatically show on the selected site), this could be a regression:

https://user-images.githubusercontent.com/17252150/167774284-0c7157bd-d357-4222-be07-fa4187a79e0a.mp4

- Noticed when `QuickStartDynamicCardsFeatureConfig` flag is turned on if a card is pinned it is still removed once the task is completed. Is that intended behavior?:

https://user-images.githubusercontent.com/17252150/167775067-4d178501-b6c5-4ede-8a64-daf4b4507e94.mp4

- Noticed when `QuickStartDynamicCardsFeatureConfig` flag is turned on, when a card is pinned, it moves up but the screen does not follow the action so it would look like the card is gone:

https://user-images.githubusercontent.com/17252150/167775477-1838ce55-6e3d-4e64-976f-eec1087417e6.mp4

I have not checked the dynamic cards observations against a production build (I'm not sure how to turn this on for production since the debug option is not there...) so thought I'd share them here so you can have a look first.

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartNoticeDetails.kt [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTask
+
+/**
+ * Static data that represents info that goes into Quick Start notices
+ */
+enum class QuickStartNoticeDetails(
+    val taskString: String,
+    val titleResId: Int,
+    val messageResId: Int
+) {
```
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartNoticeDetails.kt:15]: Nitpick: @ashiagr Thank you for converting this to Kotlin. I think adding the Resource annotation(@StringRes) to the variables here would be a good improvement. WDYT?  🤔

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartTaskDetails.kt [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+/**
+ * Static data that represents user facing description of quick start tasks.
+ */
+enum class QuickStartTaskDetails(
+    val taskString: String,
+    val titleResId: Int,
+    val subtitleResId: Int,
+    val iconResId: Int
+) {
```
Comment (AjeshRPai) [inline WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartTaskDetails.kt:18]: @ashiagr Same here, Thanks for converting this file to kotlin, Do you think we can add a resource annotation(@StringRes,@DrawableRes) improvement? 🤔

**Review by AjeshRPai — APPROVED** [review]
Thanks for tackling this task @ashiagr . Awesome job. I have reviewed the code changes and I have done regression testing on quick start. I have also tested with dynamic cards as well. Everything looks good. 👍🏼 . I have done the review for the Corresponding fluxC PR and everything looks good there as well. 

I have left two nitpicks for you to consider. As those are non blocking, I am approving the PR. Please feel free to merge if you think those are not needed.

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartNoticeDetails.kt (hunk 1 above)
Comment (ashiagr) [reply WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartNoticeDetails.kt:15]: Agree, added in 360aec2 for `QuickStartMySitePrompts`, `QuickStartNoticeDetails`, `QuickStartTaskDetails`.

**Comment by ashiagr** [comment]
Thanks, @jostnes for a detailed review!  Really appreciate it.

Please find my answers below:

> Some observations from testing:

> Noticed that user is still not guided to the correct tab (from Menu to Home) in this build. The issue looks to be fixed in https://github.com/wordpress-mobile/WordPress-Android/pull/16462. Was wondering if the fixed is included in the PR or if this is a regression?

That's correct. The PR is not yet merged to the `trunk` branch. The issue should be resolved once it gets merged.

> Noticed that there is no quick start prompt when the QuickStartDynamicCardsFeatureConfig flag is turned on (it will automatically show on the selected site), this could be a regression:

Quick Start Dynamic Cards are not enabled in production. I tested and found the same behavior in the existing code in `trunk` branch.

https://user-images.githubusercontent.com/1405144/167847964-bf9aee91-9458-4076-9d39-e01f4817ca66.mp4

> Noticed when QuickStartDynamicCardsFeatureConfig flag is turned on if a card is pinned it is still removed once the task is completed. Is that intended behavior?:

I think this is expected behavior. I tested in the existing `trunk` branch and noticed the same behaviour.

https://user-images.githubusercontent.com/1405144/167848135-a38ff72d-f159-42e9-8819-bb23c95a75d4.mp4

> Noticed when QuickStartDynamicCardsFeatureConfig flag is turned on, when a card is pinned, it moves up but the screen does not follow the action so it would look like the card is gone:

This one looks like a potential bug. I'll create an issue and link it here. Since dynamic cards are not enabled in production, I'll assign a low priority to it.

> I have not checked the dynamic cards observations against a production build (I'm not sure how to turn this on for production since the debug option is not there...) so thought I'd share them here so you can have a look first.

Thank you so much for the detailed notes. I've attached videos from the trunk branch. You can also try the APK from the previous PR: https://github.com/wordpress-mobile/WordPress-Android/pull/16473#issuecomment-1117055956 which will allow you to enable the flag in case you want to be sure. 🙂

**Comment by ashiagr** [comment]
@jostnes 

I'll go ahead and merge the PR as the issues reported are non-blocking and mostly related to QS dynamic cards which are not enabled in production and not related to changes in this PR.

Thanks once again for your inputs. 🙇‍♀️ 🙌

**Pull request merged**